### PR TITLE
produce flow code skeleton

### PR DESCRIPTION
**Pull request by kmrdhruv**
Basic skeletion for produce flow.
Review primarily entities and code flow/pattern.

### CODE REVIEW

File: entities/src/main/java/com/flipkart/varadhi/entities/InternalTopic.java [hunk 1]
```diff
@@ -67,8 +71,14 @@ public enum TopicKind {
...
     public enum ProduceStatus {
+        // topic allows produce, default status.
         Active,
-        InActive
+        // topic produce is blocked (e.g. administrative reasons)
+        Blocked,
+        // topic is being throttled currently.
+        Throttled,
+        // produce not allowed to passive replicated topic.
+        NotAllowed
```
Comment (gauravAshok) [inline entities/src/main/java/com/flipkart/varadhi/entities/InternalTopic.java:81]: rename to "Replicating".

**Review by gauravAshok — CHANGES_REQUESTED** [review]

File: common/src/main/java/com/flipkart/varadhi/exceptions/ServerNotAvailableException.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package com.flipkart.varadhi.exceptions;
+
+import lombok.experimental.StandardException;
+
+@StandardException
+public class ServerNotAvailableException extends VaradhiException {
```
[the remark this answers is not in the document]
Comment (kmrdhruv) [reply common/src/main/java/com/flipkart/varadhi/exceptions/ServerNotAvailableException.java:6]: Agree, this is a candidate for refactoring. Will take it independently.

File: common/src/main/java/com/flipkart/varadhi/utils/HeaderUtils.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+import static com.flipkart.varadhi.MessageConstants.HEADER_PREFIX;
+
+public class HeaderUtils {
+    public static Map<String, String> getVaradhiHeader(Map<String, String> headers) {
+        return headers
+                .entrySet()
+                .stream()
+                .filter(entry -> entry.getKey().toLowerCase().startsWith(HEADER_PREFIX))
+                .collect(Collectors.toMap(entry -> entry.getKey(), entry -> entry.getValue()));
```
[the remark this answers is not in the document]
Comment (kmrdhruv) [reply common/src/main/java/com/flipkart/varadhi/utils/HeaderUtils.java:14]: i think in Varadhi, we work with first value ?

File: server/src/main/java/com/flipkart/varadhi/RestVerticle.java [hunk 4]
```diff
@@ -51,31 +59,57 @@ public void start(Promise<Void> startPromise) {
...
+                vertx.executeBlocking(future -> {
+                    apiEndHandler.handle(ctx);
+                    future.complete();
+                }, resultHandler -> {
+                    if (resultHandler.succeeded()) {
+                        ctx.endRequestWithResponse(ctx.getApiResponse());
+                    } else {
+                        ctx.endRequestWithException(resultHandler.cause());
+                    }
+                });
```
[the remark this answers is not in the document]
Comment (kmrdhruv) [reply server/src/main/java/com/flipkart/varadhi/RestVerticle.java:83]: This is to return response on the original context (event loop thread).

File: messaging/src/main/java/com/flipkart/varadhi/produce/services/ProducerService.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        this.metricProvider = metricProvider;
+    }
+
+    public CompletableFuture<ProduceResult> produceToTopic(
+            Message message, String varadhiTopicName, ProduceContext produceContext
+    ) {
+        MsgProduceStatus msgProduceStatus = MsgProduceStatus.Success;
+        long produceStartTime = System.currentTimeMillis();
+        try {
+            String produceRegion = produceContext.getTopicContext().getRegion();
```
[the remark this answers is not in the document]
Comment (kmrdhruv) [reply messaging/src/main/java/com/flipkart/varadhi/produce/services/ProducerService.java:36]: correct, likely to have same value for the deployment. Currently value is derived from the config at rest layer and passed down.

File: entities/src/main/java/com/flipkart/varadhi/entities/ProduceContext.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+import io.vertx.ext.web.RoutingContext;
+import lombok.Getter;
+import lombok.Setter;
+
+import java.util.HashMap;
+import java.util.Map;
+
+@Getter
+@Setter
+public class ProduceContext {
```
[the remark this answers is not in the document]
Comment (kmrdhruv) [reply entities/src/main/java/com/flipkart/varadhi/entities/ProduceContext.java:13]: This is for transferring the rest api context to the actual produce layer, so mapped accordingly.

File: entities/src/main/java/com/flipkart/varadhi/entities/InternalTopic.java [hunk 7]
```diff
@@ -6,69 +6,60 @@
...
+                "%s.%s.%s.%s", varadhiTopicName, InternalTopic.TopicKind.Replica, replicaRegion, replicatingFromRegion);
     }
 
     public enum TopicKind {
         Main,
         Replica
     }
 
-    public enum ProduceStatus {
+    public enum TopicStatus {
```
[the remark this answers is not in the document]
Comment (kmrdhruv) [reply entities/src/main/java/com/flipkart/varadhi/entities/InternalTopic.java:54]: Not sure. At the moment it seems to contains both, Throttled(transient) vs blocked (persist-able).

File: entities/src/main/java/com/flipkart/varadhi/entities/InternalTopic.java [hunk 8]
```diff
@@ -6,69 +6,60 @@
...
-                TopicKind.Main,
-                "local",
-                null,
-                ProduceStatus.Active,
-                storageTopic
-        );
+    //internal Topic FQDN format is "<varadhiTopicName>.<kind>.<topicRegion>[.<replicatingFromRegion>]
+    //should be unique w.r.to <varadhiTopicName>
+    public static String internalMainTopicName(String varadhiTopicName, String region) {
+        return String.format("%s.%s.%s", varadhiTopicName, InternalTopic.TopicKind.Main, region);
```
[the remark this answers is not in the document]
Comment (kmrdhruv) [reply entities/src/main/java/com/flipkart/varadhi/entities/InternalTopic.java:37]: was thinking what would be better from search perspective when looking at global list ?
also this naming convention needs to be finalised w.r.to active/active, active/passive kind of scenarios.. TopicKind possibly can be avoided from it. -- let's discuss and close on this.

File: entities/src/main/java/com/flipkart/varadhi/entities/ProduceResult.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
+package com.flipkart.varadhi.entities;
+
+
+import lombok.Getter;
+
+@Getter
+public class ProduceResult {
```
[the remark this answers is not in the document]
Comment (kmrdhruv) [reply entities/src/main/java/com/flipkart/varadhi/entities/ProduceResult.java:7]: Possible candidate for removal is ProduceRestResult. It only contains messageId and that can be response itself.

**Pull request merged**